### PR TITLE
Challenge 02

### PR DESCRIPTION
**Pull request by SamuelHudec**
sorry kus mi to trvalo ešte aj dnes sa nám myčka pokazila

### CODE REVIEW

File: challenge_02/base_solution.py [hunk 1]
```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
...
         super().__init__()
-        self.W = None
+        self.e_length = embedding_length
+        self.TW = self._trainable_weights()
+
+    def _trainable_weights(self) -> Tuple[nn.Module, nn.Module, nn.Module]:
+        q = nn.Linear(self.e_length, self.e_length, bias=False)
+        k = nn.Linear(self.e_length, self.e_length, bias=False)
+        v = nn.Linear(self.e_length, self.e_length, bias=False)
+        return q, k, v
```
Comment (roman-dusek) [inline challenge_02/base_solution.py:23]: dá se to udělat v jednom nn.Linear

File: challenge_02/base_solution.py [hunk 2]
```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
...
 
-    def forward(self, x):
-        q, k, v = self.W
-        # norm_attention = F.softmax(q*k)
-        # (norm_attention * v).sum(-1)
-        return
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Scaled Dot-Product Attention
+        q, k, v = self.TW
```
Comment (roman-dusek) [inline challenge_02/base_solution.py:27]: ale tohle je hezky čistý

File: challenge_02/base_solution.py [hunk 3]
```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
...
+        w = F.softmax(w)
+        return torch.matmul(w, v(x))
 
 if __name__ == "__main__":
-    x = torch.randint(0, 500, (10, 50))  # tokenizer(text)
-    model = SelfAttention()
+    embdd_len = 50
+    x = torch.randint(0, 500, (10, embdd_len), dtype=torch.float32)  # tokenizer(text)
+    model = SelfAttention(embdd_len)
     print(model.forward(x))
```
Comment (roman-dusek) [inline challenge_02/base_solution.py]: stačí volat `model(x)`

File: challenge_02/base_solution.py [hunk 4]
```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
...
 
-    def forward(self, x):
-        q, k, v = self.W
-        # norm_attention = F.softmax(q*k)
-        # (norm_attention * v).sum(-1)
-        return
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        # Scaled Dot-Product Attention
+        q, k, v = self.TW
+        s = torch.matmul(q(x), torch.t(k(x)))
```
Comment (roman-dusek) [inline challenge_02/base_solution.py:28]: torch.t jsem ještě neviděl, standard je `tensor.T`, ale pokud máš víc dimenzí tak `tensor.transpose(-1,-2)` nebo `tensor.transpose(2,1)`

File: challenge_02/base_solution.py [hunk 5]
```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
...
+        w = nn.functional.normalize(s)
```
Comment (roman-dusek) [inline challenge_02/base_solution.py]: 1. vyměň to za `F.normalize`
2. a jinak jsi tím nejsem jistý, v originálním článku se pro normalizaci používá `/sqrt(dim_k)`, z jakého zdroje jsi to bral

File: challenge_02/base_solution.py [hunk 6]
```diff
@@ -7,17 +10,27 @@ class SelfAttention(nn.Module):
...
+        w = F.softmax(w)
+        return torch.matmul(w, v(x))
 
 if __name__ == "__main__":
-    x = torch.randint(0, 500, (10, 50))  # tokenizer(text)
-    model = SelfAttention()
+    embdd_len = 50
+    x = torch.randint(0, 500, (10, embdd_len), dtype=torch.float32)  # tokenizer(text)
```
Comment (roman-dusek) [inline challenge_02/base_solution.py:35]: spíš testuj jestli ti to funguje na [batch_size, tokens, embd_dim]

Comment (roman-dusek) [reply challenge_02/base_solution.py:35]: to se mi v minulosti vymstilo že jsem testoval pro jeden example bez dimenze batche a potom to nefungovalo 😄

File: challenge_02/multihead_solution.py [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+        # compress it together
+        to_concat = to_concat.transpose(1,2).contiguous().view(seq_len_query, self.e_length)
+
+        return out_cnct(to_concat)
+
+if __name__ == "__main__":
+    embdd_len = 512
+    heads = 8
+    x = torch.randint(0, 500, (10, embdd_len), dtype=torch.float32)
```
Comment (roman-dusek) [inline challenge_02/multihead_solution.py]: taky hoď si tam batch_size

File: challenge_02/base_solution.py (hunk 1 above)
Comment (SamuelHudec) [reply challenge_02/base_solution.py:23]: dakujem znie to zaujímavo pozriem na to :) ale pre tento prípad by som to nechal ako je len kvôli prehladnosti

File: challenge_02/base_solution.py (hunk 4 above)
Comment (SamuelHudec) [reply challenge_02/base_solution.py:28]: `t()` som našiel ako base funkciu ked chceš transponovať dvoj dimenzionalnu maticu, `torch.transpose()` používam v časti multihead

File: challenge_02/base_solution.py (hunk 5 above)
Comment (SamuelHudec) [reply challenge_02/base_solution.py]: ja som postupoval podla obrázku :D toho najslavnejšieho a tam bolo napísané normalize :D tak som našiel funkciu ktorá noramlizuje. Výhoda je že funkcia narozdiel od klasického podielu je blbomvzdorná. Ale máš pravdu podla paper tam `/sqrt(dim_k)` tak som to vymenil

File: challenge_02/base_solution.py (hunk 6 above)
Comment (SamuelHudec) [reply challenge_02/base_solution.py:35]: toto je super dik zaradu, selfattention som nechal kvôli prehladnosti ako je :) ale dalej som pridal batch size

**Comment by SamuelHudec** [comment]
uzatváram challenge a vrhnem sa hned zatjra na druhý
